src/Sylius/Bundle/CoreBundle/CatalogPromotion/Processor/CatalogPromotionRemovalProcessor.php

```diff
@@ -0,0 +1,104 @@
+<?php
+
+/*
+ * This file is part of the Sylius package.
+ *
+ * (c) Paweł Jędrzejewski
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+declare(strict_types=1);
+
+namespace Sylius\Bundle\CoreBundle\CatalogPromotion\Processor;
+
+use InvalidArgumentException;
+use SM\Factory\FactoryInterface;
+use SM\StateMachine\StateMachineInterface;
+use Sylius\Bundle\CoreBundle\CatalogPromotion\Command\RemoveInactiveCatalogPromotion;
+use Sylius\Bundle\CoreBundle\CatalogPromotion\Exception\InvalidCatalogPromotionStateException;
+use Sylius\Component\Core\Model\CatalogPromotionInterface;
+use Sylius\Component\Promotion\Event\CatalogPromotionEnded;
+use Sylius\Component\Promotion\Model\CatalogPromotionStates;
+use Sylius\Component\Promotion\Model\CatalogPromotionTransitions;
+use Sylius\Component\Promotion\Repository\CatalogPromotionRepositoryInterface;
+use Symfony\Component\Messenger\MessageBusInterface;
+
+final class CatalogPromotionRemovalProcessor implements CatalogPromotionRemovalProcessorInterface
+{
+    private StateMachineInterface $stateMachine;
+
+    public function __construct(
+        private CatalogPromotionRepositoryInterface $catalogPromotionRepository,
+        private FactoryInterface $stateMachineFactory,
+        private MessageBusInterface $commandBus,
+        private MessageBusInterface $eventBus,
+    ) {
+    }
+
+    public function removeCatalogPromotion(string $catalogPromotionCode): void
+    {
+        /** @var CatalogPromotionInterface|null $catalogPromotion */
+        $catalogPromotion = $this->getCatalogPromotion($catalogPromotionCode);
+
+        $this->initializeStateMachine($catalogPromotion);
+
+        if ($this->stateMachine->getState() === CatalogPromotionStates::STATE_INACTIVE) {
```

Comment:
The state set on catalog promotion is not always up-to-date that you are using the state machine?